devito/ir/support/utils.py

```diff
@@ -127,16 +127,17 @@ def detect_flow_directions(exprs):
             if not dimensions:
                 continue
             for d in dimensions:
-                if d.is_Sub:
-                    testdim = d.parent
-                else:
-                    testdim = d
-
+                distance = None
+                for i in d._defines:
```

Comment:
This loop is sensitive to order of iteration over a set.  That scares me a little.  Is it theoretically possible for `w.distance(r,i)` to be valid for more than one `i`?